### PR TITLE
feat: improve color consistency

### PR DESCRIPTION
**Pull request by stephenLYZ**
### SUMMARY

This PR improves the existing color consistency. The main optimization is that in a dashboard if a color scheme is set for the dashboard (either in the color scheme select or via the metadata editor), all charts in the dashboard with the same label have the same color, while other labels are generated in the normal color scheme order.

### before
image

### after
image

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/packages/superset-ui-core/src/color/CategoricalColorNamespace.ts [hunk 1]
```diff
@@ -39,8 +39,19 @@ export default class CategoricalColorNamespace {
 
   getScale(schemeId?: string) {
     const id = schemeId ?? getCategoricalSchemeRegistry().getDefaultKey() ?? '';
+    const scale = this.scales[id];
+    if (scale) {
+      return scale;
+    }
```
Comment (geido) [inline superset-frontend/packages/superset-ui-core/src/color/CategoricalColorNamespace.ts:45]: I believe this might bring a few problems back. For instance, by reusing an existing scale like this in a dashboard, the charts would change colors based on the number of charts and their rendering order. We need to avoid that. Each chart should always use its own color scale and never share an existing color scale with any other chart within the same namespace and scheme. 

This is what makes the whole implementation much harder and the core issue I was mentioning. As discussed, one way to solve this problem might be to move the logic of the color generation in the Dashboard, by pulling all the labels of all the charts and only re-apply different colors to those labels that are shared across multiple charts, rather than sharing their scale.

**Review by geido — COMMENTED** [review]
Hey @stephenLYZ thanks for working on this hard problem. Please read my comment. I am open to discuss this further.

File: superset-frontend/packages/superset-ui-core/src/color/CategoricalColorNamespace.ts (hunk 1 above)
Comment (stephenLYZ) [reply superset-frontend/packages/superset-ui-core/src/color/CategoricalColorNamespace.ts:45]: Hey @geido Thanks for the input!  This change was added temporarily and is not a major change, so I will restore it here. The major idea of this PR is to use a `default color scale` to save all the labels in the dashboard and regenerate them when the color schema is switched, as you said, but I try to keep the changes as small as possible here, there may be some problems, and I am keeping the test.

Comment (stephenLYZ) [reply superset-frontend/packages/superset-ui-core/src/color/CategoricalColorNamespace.ts]: I get all the labels by the `defaultColorScale.domain()`, then regenerate new colors when color schema changed and save them in the json metadata.

**Review by zhaoyongjie — COMMENTED** [review]
For testing purposes, used the virtual table as Dataset. We can see different `value of column` has different color:
```
SELECT * FROM 
(SELECT 'a' as color, 1 as val
UNION 
SELECT 'b', 1
UNION 
SELECT 'c', 1
UNION 
SELECT 'd', 1
UNION 
SELECT 'e', 1
UNION
SELECT 'f', 1
UNION 
SELECT 'g', 1
UNION 
SELECT 'h', 1
UNION 
SELECT 'i', 1
UNION
SELECT 'j', 1
UNION 
SELECT 'k', 1
UNION 
SELECT 'l', 1
UNION 
SELECT 'm', 1
UNION
SELECT 'n', 1
UNION 
SELECT 'o', 1
UNION 
SELECT 'p', 1
UNION 
SELECT 'q', 1
UNION
SELECT 'r', 1
UNION 
SELECT 's', 1
UNION 
SELECT 't', 1
UNION 
SELECT 'u', 1
UNION
SELECT 'v', 1
UNION
SELECT 'w', 1
UNION 
SELECT 'x', 1
UNION 
SELECT 'y', 1
UNION 
SELECT 'z', 1
) as _view
ORDER BY COLOR
```

1. Create virtual dataset in the SQLLab.
2. Create a Pie Chart 1) `count(*)` as metric 2) `color` as groupby 3) `a,b,c` in filter, save as new chart and create a dashboard.
3. Create a Pie Chart 1) `count(*)` as metric 2) `color` as groupby 3) `d,e,f` in filter, save as new chart and insert into previous dashboard.

### In this PR
image

### Current master(19daf65b5)
image

**Comment by geido** [comment]
> For testing purposes, used the virtual table as Dataset. We can see different `value of column` has different color:
> ```
> SELECT * FROM 
> (SELECT 'a' as color, 1 as val
> UNION 
> SELECT 'b', 1
> UNION 
> SELECT 'c', 1
> UNION 
> SELECT 'd', 1
> UNION 
> SELECT 'e', 1
> UNION
> SELECT 'f', 1
> UNION 
> SELECT 'g', 1
> UNION 
> SELECT 'h', 1
> UNION 
> SELECT 'i', 1
> UNION
> SELECT 'j', 1
> UNION 
> SELECT 'k', 1
> UNION 
> SELECT 'l', 1
> UNION 
> SELECT 'm', 1
> UNION
> SELECT 'n', 1
> UNION 
> SELECT 'o', 1
> UNION 
> SELECT 'p', 1
> UNION 
> SELECT 'q', 1
> UNION
> SELECT 'r', 1
> UNION 
> SELECT 's', 1
> UNION 
> SELECT 't', 1
> UNION 
> SELECT 'u', 1
> UNION
> SELECT 'v', 1
> UNION
> SELECT 'w', 1
> UNION 
> SELECT 'x', 1
> UNION 
> SELECT 'y', 1
> UNION 
> SELECT 'z', 1
> ) as _view
> ORDER BY COLOR
> ```
> 
> 1. Create virtual dataset in the SQLLab.
> 2. Create a Pie Chart 1) `count(*)` as metric 2) `color` as groupby 3) `a,b,c` in filter, save as new chart and create a dashboard.
> 3. Create a Pie Chart 1) `count(*)` as metric 2) `color` as groupby 3) `d,e,f` in filter, save as new chart and insert into previous dashboard.
> 
> ### In this PR
> image
> 
> ### Current master(19daf65b5)
> image
> 
> 

You are right @zhaoyongjie that's because this PR brings back sharing the same color instance across all the charts of a dashboard, as I mentioned in a comment above. However, @stephenLYZ said that is for testing only, so I assume it will be reverted soon.

**Comment by stephenLYZ** [comment]
> 

I have actually reverted. 😂 I believe this PR should behave as expected. Since the column(label) of the second chart is different from the first one, they may not share the same color.

**Comment by geido** [comment]
> @geido I have actually reverted. 😂 I believe this PR should behave as expected. Since the column(label) of the second chart is different from the first one, they may not share the same color.

@stephenLYZ but they have the same number of series, they should use the available colors of the color scheme from first to last, individually. What @zhaoyongjie is showing in master is the correct and expected behaviour afaik.

File: superset-frontend/plugins/plugin-chart-echarts/src/Treemap/transformProps.ts [hunk 2]
```diff
@@ -259,7 +263,7 @@ export default function transformProps(
         show: false,
       },
       itemStyle: {
-        color: CategoricalColorNamespace.getColor(),
+        color: '#1FA8C9',
```
Comment (geido) [inline superset-frontend/plugins/plugin-chart-echarts/src/Treemap/transformProps.ts:266]: Hey @stephenLYZ can you shed some light on this change for me, please?

File: superset-frontend/src/dashboard/actions/dashboardInfo.ts [hunk 3]
```diff
@@ -23,6 +23,21 @@ import { ChartConfiguration, DashboardInfo } from '../reducers/types';
 
 export const DASHBOARD_INFO_UPDATED = 'DASHBOARD_INFO_UPDATED';
 
+export function updateColorSchema(
+  metadata: any,
```
Comment (geido) [inline superset-frontend/src/dashboard/actions/dashboardInfo.ts:27]:
```suggestion
  metadata: Record<string, any>,
```

File: superset-frontend/src/explore/components/ExploreChartHeader/index.jsx [hunk 4]
```diff
@@ -157,13 +157,26 @@ export class ExploreChartHeader extends React.PureComponent {
...
+                metadata.color_scheme,
+              );
+            });
             Object.keys(labelColors).forEach(label => {
-              categoricalNamespace.setColor(label, labelColors[label]);
+              categoricalNamespace.setColor(
+                label,
+                labelColors[label],
+                metadata.color_scheme,
+              );
```
Comment (geido) [inline superset-frontend/src/explore/components/ExploreChartHeader/index.jsx:179]: What do you think about merging `sharedLabelColors` and `labelColors` in one map and then only loop through one map? Also, I think it would make sense to change the name of `labelColors` with `customLabelColors` for clarity.

File: superset/dashboards/dao.py [hunk 5]
```diff
@@ -265,6 +265,7 @@ def set_dash_metadata(  # pylint: disable=too-many-locals
         md["refresh_frequency"] = data.get("refresh_frequency", 0)
         md["color_scheme"] = data.get("color_scheme", "")
         md["label_colors"] = data.get("label_colors", {})
+        md["shared_label_colors"] = data.get("shared_label_colors", {})
```
Comment (geido) [inline superset/dashboards/dao.py:268]: The `shared_label_colors` will be presented to the user in the advanced settings of the Dashboard. I think we should hide them in the frontend so the user never sees them if possible.

**Review by geido — COMMENTED** [review]
Hey @stephenLYZ there is a lot of improvement that went into this. I am curious to see this in action. Do you think is it a good time to spin up a test env?

File: superset-frontend/plugins/plugin-chart-echarts/src/Treemap/transformProps.ts (hunk 2 above)
Comment (stephenLYZ) [reply superset-frontend/plugins/plugin-chart-echarts/src/Treemap/transformProps.ts:266]: The logic here is to display a default color when the treemap has no value, otherwise it is blank and the user may thinks it crashes. The previous way would generate a random color and give the scale an `undefined` key. That's way were wired.I think it is enough to write as one fixed color here.

File: superset-frontend/src/explore/components/ExploreChartHeader/index.jsx (hunk 4 above)
Comment (stephenLYZ) [reply superset-frontend/src/explore/components/ExploreChartHeader/index.jsx:179]: Great. I agree to change the name to `customLabelColors` but should we need to change the fields on the backend?  Or just changing the variable name here

File: superset/dashboards/dao.py (hunk 5 above)
Comment (stephenLYZ) [reply superset/dashboards/dao.py:268]: Agreed.  As well as we need to filter out this field from the user input.

**Comment by stephenLYZ** [comment]
> Hey @stephenLYZ there is a lot of improvement that went into this. I am curious to see this in action. Do you think is it a good time to spin up a test env?

I think it's okay and even though I didn't add tinyColor to resolve the color conflict, I don't think it gets in the way of the overall flow logic

File: superset-frontend/src/explore/components/ExploreChartHeader/index.jsx (hunk 4 above)
Comment (geido) [reply superset-frontend/src/explore/components/ExploreChartHeader/index.jsx:179]: If that helps bring more clarity to the backend code, I'd say yes. However, I am happy with changing it on the frontend only for now

**Comment by geido** [comment]
> > Hey @stephenLYZ there is a lot of improvement that went into this. I am curious to see this in action. Do you think is it a good time to spin up a test env?
> 
> I think it's okay and even though I didn't add tinyColor to resolve the color conflict, I don't think it gets in the way of the overall flow logic

Great. Let's make CI pass so that we can spin up the test env and start testing 🥳

**Comment by rusackas** [comment]
Upon first glance at the test environment, I see that the dashboard I looked at was... consistent! Hooray! 👏 

I also see that this solution works by effectively populating the JSON metadata of the dashboard with ALL the colors of ALL the series. I guess that works. That seems to cause a few potential issues:

1) I'm not sure if storing the metadata here might cause any potential bloat issues for the data, if there's some crazy number of series on the dashboard
2) If you go and delete the series colors from the metadata, it doesn't seem like there's any way to get it back (adding more series, changing palettes, etc).

Part of #2 is that I'm not sure what events trigger an update/refresh of the metadata. If there is such an event, I'd want to test and make sure that
1) Changing palettes (or any other action) doesn't overwrite any user-added custom colors
2) If you add a new series to two charts on a dashboard, those are also updated to match

In other words, this seems to work at first glance, but there are a few implications that make me nervous about the approach. I'm curious what @zhaoyongjie and/or @villebro might think of it.

**Comment by jinghua-qa** [comment]
Thank you for the information and i will try my best in testing, also @geido and i have a test plan for color consistency and i will use that for testing

**Comment by jinghua-qa** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_ENABLE_DND_WITH_CLICK_UX=true FEATURE_UX_BETA=true

**Comment by jinghua-qa** [comment]
I have found some issue when testing:

1, in the dashboard with label_color, if i edit color theme in meta data, the charts with label_color wont be able to change the color theme, but the charts without  label_color can, with this edit, the color is not consistent for the dashboard

https://user-images.githubusercontent.com/81597121/150499902-c5340b7c-405d-474a-b19a-32d33adb3373.mov

2, When user open a label_color chart in dashboard, and then try to save the chart, the label_color will be gone after the chart is saved.

https://user-images.githubusercontent.com/81597121/150500623-c423a4e8-ca23-4a47-8076-e518396d098c.mov

3, Under specific condition of color#, the pie chart will show  same color for connected section, this issue is filed before and still exist
https://github.com/apache/superset/issues/17443

**Comment by geido** [comment]
> Upon first glance at the test environment, I see that the dashboard I looked at was... consistent! Hooray! 👏
> 
> I also see that this solution works by effectively populating the JSON metadata of the dashboard with ALL the colors of ALL the series. I guess that works. That seems to cause a few potential issues:
> 
> 1. I'm not sure if storing the metadata here might cause any potential bloat issues for the data, if there's some crazy number of series on the dashboard
> 2. If you go and delete the series colors from `shared_label_colors`, it doesn't seem like there's any way to get it back (adding more series, changing palettes, etc).
> 3. Normally people would add custom colors to `label_colors`, but it's now possible (and easy and discoverable) to edit the `shared_label_colors` object instead. This may lead to problems with these colors being overwritten when changing color palette.
> 

I totally agree with this. I think we should hide the object from the user as discussed previously. @stephenLYZ this shouldn't be a big problem. We should be able to delete the object from the editor before its presented to the user.

As for what concerns storing ALL labels, I think we agreed to only store the labels that are shared across different charts and not all of them. This might require a bit more logic in the code but it's important to avoid flooding the metadata.

> There are more cases I didn't test. For example, If you add a new series to two charts on a dashboard, hopefully those are also updated to match other existing charts. We should capture all the test cases for either manual or (better yet) automated tests to make sure it all works. Pinging @jinghua-qa to make her aware of the complexity here.
> 
> In other words, this seems to work at first glance, but there are a few implications that make me nervous about the approach. I'm curious what @zhaoyongjie and/or @villebro might think of it.
> 
> One solution to much of the above might be to forgo putting the `shared_label_colors` in the (user editable) json metadata, and instead using the dashboard's new key/value endpoint. Then it's in the database, and (hopefully?) could be edited repeatedly without the user even knowing it exists. I think it could still be accessed as needed by Explore, too, but I'm not sure.

I think using the new key/value endpoint is a good idea. We could make sure the approach works correctly by using the metadata and then moving away from it in phase 2.

**Comment by geido** [comment]
@stephenLYZ I also made a little bit of testing, below are some thoughts/findings:

- When I remove the charts that share labels, I would expect that the `shared_label_colors` will get cleaned up from all of the labels that are not shared or that do not exist anymore in the context of the Dashboard. Imagine someone working on a large Dashboard and adding/removing charts multiple times, the `shared_label_colors` will eventually get very large with data that is not really needed.
- When changing the color scheme from the Dashboard properties -> Color scheme, I saw that the `shared_label_colors` was wiped out and not re-generated, even though I have charts sharing labels. I was expecting that to be re-generated with the updated colors from the chosen scheme. It looks like this only happens when you remove or add a new chart in the Dashboard and then you change the color scheme. Watch the video below

https://user-images.githubusercontent.com/60598000/150560405-1d169b24-686f-4204-831e-252f3b4dc42f.mp4

- When you add charts to the Dashboard, it looks like the `shared_label_colors` is not updated. It only gets updated when you change the color scheme. Also, it appears that the previous labels that were there already got deleted while they shouldn't.

https://user-images.githubusercontent.com/60598000/150562945-c3d0f2ab-6c48-41b5-a376-4f13dee72a9c.mp4

- I deleted some charts and cleaned up the `shared_label_colors` for testing. Then I changed the color scheme. The `shared_label_colors`  is still getting the labels also of the charts that I deleted. I think it's not getting the most updated state of the Dashboard. It gets the correct state only when hard-refreshing the Dashboard page.

I'll be back to testing once these things are tackled as I don't want to put too much on the plate right now. Let's jump on a call if anything isn't clear or if you need any help sorting these out!

**Comment by stephenLYZ** [comment]
> I have found some issue when testing:
> 
> 1, in the dashboard with label_color, if i edit color theme in meta data, the charts with label_color wont be able to change the color theme, but the charts without label_color can, with this edit, the color is not consistent for the dashboard
> 
>  Screen.Recording.2022-01-21.at.12.52.32.AM.mov 
> 2, When user open a label_color chart in dashboard, and then try to save the chart, the label_color will be gone after the chart is saved.
> 
>  Screen.Recording.2022-01-21.at.12.36.14.AM.mov 
> 3, Under specific condition of color#, the pie chart will show same color for connected section, this issue is filed before and still exist #17443

1. fixed.
2 & 3. Currently the master has the same problem, I will open another PR to solve.

**Comment by stephenLYZ** [comment]
@geido @rusackas For now I will hide the shared_label_color field([here](https://github.com/apache/superset/pull/17792/files#diff-bbab1592904855ed63c204cedb8c14efc9eed397dac25756ac32f1cca4843518R173)) and filter out the shared_label_color entered by the user([here](https://github.com/apache/superset/pull/17792/files#diff-bbab1592904855ed63c204cedb8c14efc9eed397dac25756ac32f1cca4843518R297)).  And there are only two ways for users to update shared_label_color:
1. update color scheme, whether by color scheme select or json metadata editor:

https://user-images.githubusercontent.com/11830681/153238011-91ec5b2a-a6e3-4f20-9165-1330be04923f.mov

2. add chart or delete chart:


https://user-images.githubusercontent.com/11830681/153238370-0e9d9b71-295a-4eba-a522-c545d088f84d.mov

**Comment by geido** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_ENABLE_DND_WITH_CLICK_UX=true FEATURE_UX_BETA=true

**Comment by geido** [comment]
Hey @stephenLYZ some comments after additional testing.

I had three charts sharing labels and the colors look right with using the Lyft Colors scheme. I removed two of the charts, leaving just one. Then I added back another chart that shared labels with the remaining one and the colors are not the same any longer. Check the screenshot.

Even after changing the color scheme the label colors still look incorrect for the first chart. It's like at some point it got ignored.

You can see this behavior in Dashboard http://54.202.2.94:8080/superset/dashboard/12 that I'll leave untouched on the test env for you to reproduce the problem.

Screen Shot 2022-02-11 at 15 44 36

In a Dashboard where no color scheme is used, I see that the same labels do not share the same colors. I am not sure if this is the wanted behavior but probably worth checking

Screen Shot 2022-02-11 at 15 54 27

For some reason, in that same Dashboard without color scheme, after applying the Lyft Colors color scheme, the label colors of the first chart are different from the others. I was under the assumption that with changing the color scheme the label colors would be regenerated, but it looks like this does not happen for the first chart for some reasons. You can check that behavior here http://54.202.2.94:8080/superset/dashboard/11. I think it's the same problem as the first point above.

Screen Shot 2022-02-11 at 15 58 16

**Comment by jinghua-qa** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_ENABLE_DND_WITH_CLICK_UX=true FEATURE_UX_BETA=true

**Comment by geido** [comment]
Hello @stephenLYZ some more comments after another round of testing:

- In the sample Dashboard "World Bank's Data" I noticed different label colors. I am showing one of the cases in the screenshot. Some other charts seem to work correctly instead within the same Dashboard. FYI this Dashboard has a color scheme and no custom label colors.

Screen Shot 2022-02-16 at 19 19 20

This looks so nearly done! I am going to do another code related review as I think manual testing was pretty awesome so far. Great job!

File: superset-frontend/src/dashboard/actions/dashboardState.js [hunk 6]
```diff
@@ -474,6 +480,21 @@ export function addSliceToDashboard(id, component) {
       if (selectedSlice && selectedSlice.viz_type === 'filter_box') {
         dispatch(addFilter(id, component, selectedSlice.form_data));
       }
+
+      setTimeout(() => {
```
Comment (geido) [inline superset-frontend/src/dashboard/actions/dashboardState.js:484]: You told me about this timeout but the more I think about it the more I am concerned that it might fail in a few instances, especially in large dashboards. We should probably look into how the dashboard should signal that all charts have finished querying. I think this is important for other use cases as well. @rusackas what do you think?

Comment (rusackas) [reply superset-frontend/src/dashboard/actions/dashboardState.js:484]: Agreed. I'm not sure what the 500ms is based on, but it feels arbitrary compared to what's likely to happen in reality. I'm hopeful that there's something we can `await` here. I'm not sure exactly what yet... will need to dig a bit deeper.

**Comment by jinghua-qa** [comment]
I found an issue that when user customize the color for a label in a chart, then user open the chart in explore and save, the color order got refreshed and  the customize color label did not stay,  is this expected?

https://user-images.githubusercontent.com/81597121/154443187-4d434edc-3fd0-4676-8437-8a2498585027.mov

File: superset-frontend/packages/superset-ui-core/src/color/SharedLabelColorSingleton.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    // save all shared label value
+    this.values = [];
+    // { value1: [sliceId1, sliceId2], value2: [sliceId], ...}
+    this.valueSliceMap = {};
+  }
+
+  getColorMap(colorNamespace?: string, colorScheme?: string) {
+    if (colorScheme) {
+      const categoricalNamespace = CategoricalColorNamespace.getNamespace(
+        colorNamespace || '',
```
[the remark this answers is not in the document]
Comment (stephenLYZ) [reply superset-frontend/packages/superset-ui-core/src/color/SharedLabelColorSingleton.ts:38]: Good Catch! The or clause may not be needed here, as the `colorNamespace` is `undefined` at this point and will take `GLOBAL` by default

**Review by jinghua-qa — APPROVED** [review]
Did manual QA test for color consistency on tier 1 charts and it looks good to me. There is one issue for the customize color for specific label, but it is an existing issue in current master, would love to see it fix in follow up PR

**Comment by stephenLYZ** [comment]
> Hello @stephenLYZ some more comments after another round of testing:
> 
> * In the sample Dashboard "World Bank's Data" I noticed different label colors. I am showing one of the cases in the screenshot. Some other charts seem to work correctly instead within the same Dashboard. FYI this Dashboard has a color scheme and no custom label colors.
> 
> Screen Shot 2022-02-16 at 19 19 20
> 
> This looks so nearly done! I am going to do another code-related review as I think manual testing was pretty awesome so far. Great job! 🚀 🧑‍🚀

Yes. Here need to add color consistency to the map charts.
image

File: superset-frontend/src/dashboard/actions/dashboardState.js (hunk 6 above)
Comment (stephenLYZ) [reply superset-frontend/src/dashboard/actions/dashboardState.js]: Yes. In this case it is not only necessary to wait for the chart data from the server, but the exact timing is after the chart transform props, after the colon called. So I'm using `postTransformProps` to handle this (an api that seems to be rarely used, please let me know if this makes sense). See the video, I simulated a weak network environment, but the colors still updated after adding the chart.

https://user-images.githubusercontent.com/11830681/154857051-d76764cb-95cf-4bfd-b2dc-233bc95d9db3.mov

**Comment by geido** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_ENABLE_DND_WITH_CLICK_UX=true FEATURE_UX_BETA=true

**Comment by geido** [comment]
Hey @stephenLYZ I think I found a rather impactful issue. 
Basically, when you change Dashboard, the color of the labels is persisted and applied also to other Dashboards. You can see that those get correctly removed only when hard-refreshing the page.

https://user-images.githubusercontent.com/60598000/155148263-29f8336e-1c20-4184-b6e6-e78f331a4677.mp4

**Comment by geido** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_ENABLE_DND_WITH_CLICK_UX=true FEATURE_UX_BETA=true

**Comment by geido** [comment]
Hey @stephenLYZ I found that two labels within the same chart share exactly the same color. This reminded me of the color conflict resolution requirements that I think we skipped to prioritize testing looking at the threads above. What do you think?

Screen Shot 2022-02-23 at 12 53 13

**Comment by stephenLYZ** [comment]
Discussion result with @geido :
using `tinycolor.analogous `to generate similar colors for each color scheme, for example, if our color scheme is [A, B, C] then the generated colors are [A1, B1, C1, A2, B2, C2 ....]. Here the number of color labels is always less than or equal to the generated list of colors, in order to prevent getting the same colors.

**Comment by geido** [comment]
> 

Just for the sake of clarity, I am adding a bit more info here.

Let's say that we have Chart A and Chart B that share two labels. The same color for both labels would be set, but now Chart B has an adjacent label that also uses the same color. That's the main color conflict we need to resolve here.

The solution that Stephen is proposing above, generates new colors for the `shared_label_colors` using the `analogous` utility. That would assure (in most cases) that two adjacent labels would not use exactly the same color.

I think the solution works on paper, but I'd like to hear feedback from @rusackas @zhaoyongjie @jinghua-qa

**Comment by geido** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_ENABLE_DND_WITH_CLICK_UX=true FEATURE_UX_BETA=true
